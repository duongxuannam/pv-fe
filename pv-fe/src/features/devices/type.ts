export interface Device {
  id: string
  name: string
  last_update: string | Date
  location: string
  status: string
  current_temperature: string | number
}
