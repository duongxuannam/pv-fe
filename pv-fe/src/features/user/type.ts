export interface User {
  id: string
  name: string
  title: string
}
