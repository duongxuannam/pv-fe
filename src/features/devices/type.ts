interface Device {
    id: string;
    name: string;
    last_update: string;
    location: string;
    status: string;
    current_temperature: string | number;
}