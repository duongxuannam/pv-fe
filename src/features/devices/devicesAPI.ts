
import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/query/react'
import { baseURL } from '../../utils/apiRequest'

export const fetchDevices = async () => {
  try {
    const response = await fetch(
      "http://localhost:3000/devices"
    )
    const data = await response.json()
    return data
  } catch (error) {
    return error
  }

}

export const devicesApi = createApi({
  reducerPath: 'devicesApi',
  baseQuery: fetchBaseQuery({ baseUrl: baseURL }),
  endpoints: (builder) => ({
    getDevices: builder.query<Device[], void | null>({
      query: () => `devices`,
    }),
  }),
})


export const { useGetDevicesQuery } = devicesApi

