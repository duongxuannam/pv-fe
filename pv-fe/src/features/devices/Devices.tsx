import useWindowFocus from "use-window-focus"
import { useGetDevicesQuery } from "./devicesAPI"
import { Device as DeviceType } from "./type"
import Device from "../../components/Device"
import Loading from "../../components/Loading"

const IS_RUNNING_TEST = process.env.NODE_ENV === "test"

function Devices() {
  const isWindowFocused = useWindowFocus()
  const { data, error, isLoading, refetch } = useGetDevicesQuery(null, {
    pollingInterval: 3000,
    skip: IS_RUNNING_TEST ? false : !isWindowFocused,
  })
  return (
    <>
      <div className="flex justify-end mt-10 mr-10">
        <button
          onClick={refetch}
          className=" text-end bg-transparent hover:bg-blue-500 text-blue-700 font-semibold hover:text-white py-2 px-4 border border-blue-500 hover:border-transparent rounded"
        >
          Refresh
        </button>
      </div>
      <div className="m-10">
        {!!error && (
          <span
            data-testid="error_test"
            className="text-rose-600 text-xl font-medium mx-0"
          >
            Có lỗi xãy ra
          </span>
        )}
        <div className="grid gap-5 2 grid-cols-1 md:grid-cols-2 lg:grid-cols-4 xl:grid-cols-4">
          {isLoading &&
            Array(8)
              .fill(1)
              .map((_, index) => {
                return <Loading key={index} />
              })}

          {data &&
            !isLoading &&
            data.map((item: DeviceType) => {
              const date = new Date(item.last_update)
              return (
                <Device
                  name={item.name}
                  key={item.id}
                  date={date}
                  id={item.id}
                  location={item.location}
                  current_temperature={item.current_temperature}
                  last_update={item.last_update}
                  status={item.status}
                />
              )
            })}
        </div>
      </div>
    </>
  )
}

export default Devices
