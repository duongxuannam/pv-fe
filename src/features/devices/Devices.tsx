import useWindowFocus from 'use-window-focus';
import { useGetDevicesQuery } from "./devicesAPI"

function Devices() {
    const isWindowFocused = useWindowFocus()
    const { data, error, isLoading, refetch } = useGetDevicesQuery(null, {
        pollingInterval: 3000, skip: !isWindowFocused
    })

    return (
        <>

            <div className="flex justify-end mt-10 mr-10">
                <button
                    onClick={refetch}
                    className=" text-end bg-transparent hover:bg-blue-500 text-blue-700 font-semibold hover:text-white py-2 px-4 border border-blue-500 hover:border-transparent rounded">
                    Refresh
                </button>
            </div>
            <div className="m-10">

                <div className="grid  gap-5 2 grid-cols-1 md:grid-cols-2 lg:grid-cols-4">
                    {
                        isLoading && (
                            Array(8).fill(1).map((item, index) => {
                                return (
                                    <div key={index} className="flex w-full max-w-md mx-auto overflow-hidden  rounded-lg shadow-lg bg-gray-300 animate-pulse dark:bg-gray-800">
                                        <div className="h-60 w-max bg-gray-300 dark:bg-gray-600"></div>
                                    </div>
                                )
                            })
                        )
                    }

                    {data && !isLoading && (
                        data.map((item: Device) => {
                            const date = new Date(item.last_update)

                            return (
                                <div key={item.id} className="p-5 m-1 border-solid border-2 border-indigo-600 flex flex-col items-start gap-1">
                                    <span className="block">
                                        <b>ID:</b>  {item.id}
                                    </span>
                                    <span>
                                        <b>Name:</b> {item.name}
                                    </span>
                                    <span>
                                        <b>Location:</b> {item.location}
                                    </span>
                                    <span>
                                        <b>Status:</b> {item.status}
                                    </span>
                                    <span>
                                        <b>Temperature:</b>  <span className={Number(item.current_temperature) > 30 ? 'text-red-500 font-bold' : 'text-green-400'}  >
                                            {item.current_temperature}   </span>
                                    </span>
                                    <span>
                                        <b>Last update:</b>{date.toLocaleTimeString()} - {date.toLocaleDateString()}
                                    </span>
                                </div>
                            )
                        })
                    )}
                </div>
            </div >
        </>

    )
}

export default Devices
