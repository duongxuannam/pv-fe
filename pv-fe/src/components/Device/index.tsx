import React from "react"
import { Device as DeviceType } from "../../features/devices/type"
import { useAppSelector } from "../../app/hooks"

interface Props extends Partial<DeviceType> {
  date: Date
}

const Device = ({
  id,
  name,
  location,
  status,
  current_temperature,
  date,
}: Props) => {
  const role = useAppSelector(selectUserRole())
  role
  return (
    <div
      key={id}
      className="rounded-md shadow-lg ring-1 p-5 m-1 border-solid border border-zinc-500 flex flex-col items-start gap-1 "
    >
      <span className="block" data-testid="id">
        <b>ID:</b> <span>{id}</span>
      </span>
      <span data-testid="name">
        <b>Name:</b> {name}
      </span>
      <span data-testid="location">
        <b>Location:</b> {location}
      </span>
      {role.admin && (
        <span data-testid="status">
          <b>Status:</b> {status}
        </span>
      )}
      <span>
        <b>Temperature:</b>{" "}
        <span
          className={
            Number(current_temperature) > 30
              ? "text-red-500 font-bold"
              : "text-green-400"
          }
        >
          {current_temperature}{" "}
        </span>
      </span>
      <span>
        <b>Last update:</b>
        {date.toLocaleTimeString()} - {date.toLocaleDateString()}
      </span>
    </div>
  )
}

export default Device
