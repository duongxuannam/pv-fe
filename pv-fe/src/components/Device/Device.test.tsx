import { render, screen } from "@testing-library/react"
import Device from "."
import { Device as DeviceType } from "../../features/devices/type"

const device: DeviceType = {
  id: "1",
  current_temperature: "20",
  last_update: new Date("08/01/2023 12:37:53 AM"),
  location: "Binh Duong",
  name: "Thiet bi 1",
  status: "connected",
}

test("loads and displays ui", async () => {
  // ARRANGE
  const view = render(
    <Device
      id={device.id}
      current_temperature={device.current_temperature}
      location={device.location}
      date={device.last_update as Date}
      name={device.name}
      status={device.status}
      key={device.id}
    />,
  )

  // ACT
  // do something if need action

  // ASSERT
  expect(view).toMatchSnapshot()
  expect(screen.getByTestId("id")).toHaveTextContent(`ID: ${device.id}`)
  expect(screen.getByTestId("name")).toHaveTextContent(`Name: ${device.name}`)
  expect(screen.getByTestId("location")).toHaveTextContent(
    `Location: ${device.location}`,
  )
  expect(screen.getByTestId("status")).toHaveTextContent(
    `Status: ${device.status}`,
  )
})
