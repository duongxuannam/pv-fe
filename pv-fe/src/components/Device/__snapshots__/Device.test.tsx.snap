// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`loads and displays ui 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="rounded-md shadow-lg ring-1 p-5 m-1 border-solid border border-zinc-500 flex flex-col items-start gap-1 "
      >
        <span
          class="block"
          data-testid="id"
        >
          <b>
            ID:
          </b>
           
          <span>
            1
          </span>
        </span>
        <span
          data-testid="name"
        >
          <b>
            Name:
          </b>
           
          Thiet bi 1
        </span>
        <span
          data-testid="location"
        >
          <b>
            Location:
          </b>
           
          Binh Duong
        </span>
        <span
          data-testid="status"
        >
          <b>
            Status:
          </b>
           
          connected
        </span>
        <span>
          <b>
            Temperature:
          </b>
           
          <span
            class="text-green-400"
          >
            20
             
          </span>
        </span>
        <span>
          <b>
            Last update:
          </b>
          12:37:53 AM
           - 
          8/1/2023
        </span>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="rounded-md shadow-lg ring-1 p-5 m-1 border-solid border border-zinc-500 flex flex-col items-start gap-1 "
    >
      <span
        class="block"
        data-testid="id"
      >
        <b>
          ID:
        </b>
         
        <span>
          1
        </span>
      </span>
      <span
        data-testid="name"
      >
        <b>
          Name:
        </b>
         
        Thiet bi 1
      </span>
      <span
        data-testid="location"
      >
        <b>
          Location:
        </b>
         
        Binh Duong
      </span>
      <span
        data-testid="status"
      >
        <b>
          Status:
        </b>
         
        connected
      </span>
      <span>
        <b>
          Temperature:
        </b>
         
        <span
          class="text-green-400"
        >
          20
           
        </span>
      </span>
      <span>
        <b>
          Last update:
        </b>
        12:37:53 AM
         - 
        8/1/2023
      </span>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
